link_shortener/form_routes.py

```diff
@@ -219,71 +219,41 @@ async def update_link_save(request, user, status, link_id):
+                fresh_salt = os.urandom(32)
+                password = hashlib.pbkdf2_hmac(
+                    'sha256',
+                    form.password.data.encode('utf-8'),
+                    fresh_salt,
+                    100000
                 )
+                await conn.execute(link_update.values(
+                    url=form.url.data,
+                    switch_date=form.switch_date.data,
+                    password=password
+                ))
+                link_query = await conn.execute(table.select().where(
+                    table.columns['id'] == link_id
+                ))
                 link_data = await link_query.fetchone()
-                if link_data.password:
-                    fresh_salt = os.urandom(32)
-                    password = hashlib.pbkdf2_hmac(
-                        'sha256',
-                        form.password.data.encode('utf-8'),
-                        fresh_salt,
-                        100000
-                    )
-                    await conn.execute(
-                        table.update().where(
-                            table.columns['id'] == link_id
-                        ).values(
-                            url=form.url.data,
-                            password=password,
-                            switch_date=form.switch_date.data
-                        )
-                    )
-                    await conn.execute(
-                        salts.update().where(
-                            salts.columns['identifier'] == link_data.identifier
-                        ).values(
-                            salt=fresh_salt
-                        )
-                    )
 
+                if link_data.password:
+                    await conn.execute(salts.update().where(
+                        salts.columns['identifier'] == link_data.identifier
+                    ).values(salt=fresh_salt))
                 else:
-                    salt = os.urandom(32)
-                    password = hashlib.pbkdf2_hmac(
-                        'sha256',
-                        form.password.data.encode('utf-8'),
-                        salt,
-                        100000
-                    )
-                    await conn.execute(
-                        table.update().where(
-                            table.columns['id'] == link_id
-                        ).values(
-                            url=form.url.data,
-                            password=password,
-                            switch_date=form.switch_date.data
-                        )
-                    )
-                    await conn.execute(
-                        salts.insert().values(
+                    await conn.execute(salts.insert().values(
                             identifier=link_data.identifier,
-                            salt=salt
-                        )
-                    )
+                            salt=fresh_salt
+                    ))
```

Comment:
Heh, took me some time to find this out... This code is never reached. Guess why!... :)

I'll help you... You set the password and then you query the link data. In this order, the password is always set. You probably want to query the (old) link data before you set the password. At this moment, if you reset the password of a link and then set a new password, `UPDATE` of a salt is executed instead of the expected `INSERT`. Thus no salt is being set. This manifests when you try to follow the link. If you set a correct password, no salt is found, an exception is raised and a massage that the link does not exist is displayed to a user.

Reply:
Oh, I must've missed that during click testing. I need a robot 😄.

Thank you, should be working now.

Reply:
...or end to end tests... :trollface: